lib/assert.js

```diff
@@ -369,9 +374,114 @@ function isSpecial(obj) {
 }
 
 const typesToCallDeepStrictEqualWith = [
-  isKeyObject, isWeakSet, isWeakMap, Buffer.isBuffer,
+  isKeyObject, isWeakSet, isWeakMap, Buffer.isBuffer, isSharedArrayBuffer,
 ];
 
+function compareMaps(actual, expected, comparedObjects) {
+  if (actual.size !== expected.size) {
+    return false;
+  }
+  const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+
+  comparedObjects ??= new SafeWeakSet();
+
+  for (const { 0: key, 1: val } of safeIterator) {
+    if (!MapPrototypeHas(expected, key)) {
+      return false;
+    }
+    if (!compareBranch(val, MapPrototypeGet(expected, key), comparedObjects)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function compareArrayBuffers(actual, expected) {
+  const actualView = ArrayBufferIsView(actual) ? actual : new Uint8Array(actual);
+  const expectedView = ArrayBufferIsView(expected) ? expected : new Uint8Array(expected);
+
+  if (ObjectPrototypeToString(actualView) !== ObjectPrototypeToString(expectedView)) {
+    return false;
+  }
+
+  // Compare the lengths of the views (not just byte length, but actual element count)
+  if (expectedView.length > actualView.length) {
+    return false;
+  }
+
+  for (let i = 0; i < expectedView.length; i++) {
```

Comment:
It feels quite arbitrary to consider that comparing an `ArrayBuffer` and a `Uint8Array` would be accepted, but not any other class. I think we would benefit from separating the different cases more, and we can also avoid calling some getters.
```suggestion
  const areArrayBuffers = !ArrayBufferIsView(actual);
  const areDataViews = !areArrayBuffers && isDataView(actual);
  
  let actualView, expectedView, expectedViewLength;
  if (areArrayBuffers) {
    if (ArrayBufferIsView(expected)) return false;
    expectedViewLength = ArrayBufferPrototypeGetByteLength(expected)
    if (ArrayBufferPrototypeGetByteLength(actual) > expectedViewLength) return false;
    actualView = new Uint8Array(actual);
    expectedView = new Uin8Array(expected);
  } else if (areDataViews) {
    if (!isDataView(expected)) return false;
    const actualByteLength = DataViewPrototypeGetByteLength(actual);
    expectedViewLength = DataViewPrototypeGetByteLength(expected)
    if (actualByteLength > expectedViewLength) return false;
    actualView = new Unit8Array(
      DataViewPrototypeGetBuffer(actual),
      DataViewPrototypeGetByteOffset(actual),
      actualByteLength,
    );
    expectedView = new Unit8Array(
      DataViewPrototypeGetBuffer(expected),
      DataViewPrototypeGetByteOffset(expected),
      expectedViewLength,
    );
  } else {
    if (ObjectPrototypeToString(actual) !== ObjectPrototypeToString(expected)) return false;
    actualView = actual;
    expectedView = expected;
    expectedViewLength = TypedArrayPrototypeGetLength(expected);
    if (TypedArrayPrototypeGetLength(actual) > expectedViewLength) return false;
  }

  for (let i = 0; i < expectedViewLength; i++) {
```